hircluster.c

```diff
@@ -3781,6 +3701,132 @@ static void redisClusterAsyncCallback(redisAsyncContext *ac, void *r,
     cluster_async_data_free(cad);
 }
 
+/* Reply callback function for CLUSTER SLOTS */
+void clusterSlotsReplyCallback(redisAsyncContext *ac, void *r, void *privdata) {
+    UNUSED(ac);
+    redisClusterAsyncContext *acc = (redisClusterAsyncContext *)privdata;
+    redisClusterContext *cc = acc->cc;
+    acc->update_route_time = hi_usec_now() + SLOTMAP_UPDATE_THROTTLE_USEC;
+
+    if (r == NULL) {
+        /* Ignore failures, next random node is hopefully working */
+        return;
+    }
```

Comment:
Should we set update_route_time even if r == NULL?

Shouldn't we trigger another slot update with a different node here?

Reply:
This was a bit simplified as a start.
Hiredis calls the callback-function with a NULL reply when either the command has timedout, or when there is connection errors.

The scenario I thought of:
If a slot map update don't find any connected node it will take a random node and start an async connect, it will also send/give the CLUSTER SLOTS command to hiredis for it to send it when the connection is up.
If the connect fails this callback is called with a NULL reply, due to the outstanding CLUSTER SLOTS command.
If all redis nodes are gone (like in the issue we have) I was afraid slotmap updates would go into a busy-loop.

One option could be to only trigger a new slotupdate to one of connected nodes, if that exists.
This would speedup updates for timeout cases, or temporary node issues.
..or some bookkeeping to avoid busy-looping cases..

Reply:
I think we want a loop until we have a slot map, but wait one second between each attempt. (That's what we do in ered.)

Reply:
Waiting a second is the problem here, hiredis owns the eventloop and its timers.
We need to add a interface and functionality to each event-adapter to be able to kick-off timers in each eventsystem (or have hiredis expose something).

The modularity and being able to choose an own eventsystem its a bit annoying :-)
We have users that has implemented an own eventsystem as well, so updating existing adapters might not be enough for everybody.

Reply:
The PR is now updated to retry until all nodes has been attempted (by using a lastConnectionAttempt and throttling).
Since there are no timerhandling available yet to hiredis-cluster the first slotmap update attempt is triggered by the user via a command, the retries are triggered by the NULL-reply callback.